[WIP] feat: property template string

mierak opened the pull request:
This is a working implementation of templating as discussed here https://github.com/mierak/rmpc/issues/263. I am mostly looking at opinions/suggestions at this point. Tagging you because you have shown interest in this @roxwize.

Currently it works only for `Property` panes in the layout and you have to provide `format` instead of `content`.

It allows to transform for example this: 
```ron
(
    size: "23",
    pane: Pane(Property(
        content: [
            (kind: Property(Status(Elapsed))), 
            (kind: Text(" / ")), 
            (kind: Property(Status(Duration))), 
            (kind: Group([
                (kind: Text(" (")), 
                (kind: Property(Status(Bitrate))), 
                (kind: Text(" kbps)")),
            ])),
        ], 
        align: Left,
    )),
),
```

into this:
```ron
(
    size: "23",
    pane: Pane(Property(
        format: "$elapsed$' / '$duration$[$' ('$bitrate$' kpbs)']",
        align: Left,
    )),
),
```

These definitions are equivalent.

Groups are 

Some syntax remarks:
* `Group` kind are specified as a dollar sign and square brackets`$[ ... ]`
* `Text` kind is a dollar sign and either single or double quotes: `$'this is raw text'`, `$"also raw text"`, `$"escaping \"works too\""`
* `Sticker` is currently `$sticker(name: "playCount")`'
* other properties are similar: `$consume(onLabel: "on", offLabel: "off")`
* `default`(fallback value) is delimited by `|`: `$title|$'No song'`
* styles are in square braces following a property: `$title{fg: yellow, bg: black, mods: bold}|$'No song'{fg: red, bg: black}`

You can see some tests in parser.rs to see more examples.

mierak commented:
fwiw, I have replaced 1:1 every single property occurence in my theme (close to the default one) and ran into no issues

<details>
<summary>Click to see diff</summary>

```diff
diff --git a/dot_config/rmpc/themes/def.ron b/dot_config/rmpc/themes/def.ron
index a36983f..355c22e 100644
--- a/dot_config/rmpc/themes/def.ron
+++ b/dot_config/rmpc/themes/def.ron
@@ -28,31 +28,16 @@
                                 panes: [
                                     (
                                         size: "23",
-                                        pane: Pane(Property(
-                                            content: [
-                                                (kind: Text("["), style: (fg: "yellow", modifiers: "Bold")),
-                                                (kind: Property(Status(StateV2())), style: (fg: "yellow", modifiers: "Bold")),
-                                                (kind: Text("]"), style: (fg: "yellow", modifiers: "Bold")),
-                                            ], align: Left,
-                                        )),
+                                        pane: Pane(Property(format: "$'['{fg: yellow, mods: bold}$state(){fg: yellow, mods: bold}$']'{fg: yellow, mods: bold}", align: Left)),
                                     ),
                                     (
                                         size: "100%",
                                         borders: "LEFT | RIGHT",
-                                        pane: Pane(Property(
-                                            content: [
-                                                (kind: Property(Song(Filename)), style: (modifiers: "Bold"),
-                                                    default: (kind: Text("No Song"), style: (modifiers: "Bold")))
-                                            ], align: Center, scroll_speed: 1
-                                        )),
+                                        pane: Pane(Property(format: "$s:title{mods: bold}|$'No Song'{mods: bold}", align: Center, scroll_speed: 1)),
                                     ),
                                     (
                                         size: "23",
-                                        pane: Pane(Property(
-                                            content: [
-                                                (kind: Property(Widget(Volume)), style: (fg: "blue", modifiers: "Bold"))
-                                            ], align: Right
-                                        )),
+                                        pane: Pane(Property( format: "$w:volume{fg: blue, mods: bold}", align: Right)),
                                     ),
                                 ],
                             ),
@@ -65,63 +50,22 @@
                                 panes: [
                                     (
                                         size: "23",
-                                        pane: Pane(Property(
-                                            content: [
-                                                (kind: Property(Status(Elapsed))), 
-                                                (kind: Text(" / ")), 
-                                                (kind: Property(Status(Duration))), 
-                                                (kind: Group([
-                                                    (kind: Text(" (")), 
-                                                    (kind: Property(Status(Bitrate))), 
-                                                    (kind: Text(" kbps)")),
-                                                ])),
-                                            ], align: Left,
-                                        )),
+                                        pane: Pane(Property(format: "$elapsed$' / '$duration$[$' ('$bitrate$' kpbs)']", align: Left)),
                                     ),
                                     (
                                         size: "100%",
                                         borders: "LEFT | RIGHT",
-                                        pane: Pane(Property(
-                                            content: [
-                                                (kind: Property(Song(Artist)), style: (fg: "yellow", modifiers: "Bold"),
-                                                    default: (kind: Text("Unknown"), style: (fg: "yellow", modifiers: "Bold"))),
-                                                (kind: Text(" - ")),
-                                                (kind: Property(Song(Album)), default: (kind: Text("Unknown Album")))
-                                            ], align: Center, scroll_speed: 2
-                                        )),
+                                        pane: Pane(Property(format: "$s:artist{fg: yellow, mods: bold}|$'Unknown Artist'{fg: yellow, mods: bold}$' - '$s:album{fg: yellow, mods: bold}|$'Unknown Album'", align: Center, scroll_speed: 2)),
                                     ),
                                     (
                                         size: "23",
-                                        pane: Pane(Property(content: [
-                                            (kind: Property(Status(RepeatV2(
-                                                on_label: " ",
-                                                off_label: " ",
-                                                on_style: (fg: "yellow", modifiers: "Bold"),
-                                                off_style: (fg: "blue", modifiers: "Dim"),
-                                            )))),
-                                            (kind: Property(Status(RandomV2(
-                                                on_label: " ",
-                                                off_label: " ",
-                                                on_style: (fg: "yellow", modifiers: "Bold"),
-                                                off_style: (fg: "blue", modifiers: "Dim"),
-                                            )))),
-                                            (kind: Property(Status(SingleV2(
-                                                on_label: "󰑘 ",
-                                                off_label: "󰑘 ",
-                                                oneshot_label: "󰑘 ",
-                                                on_style: (fg: "yellow", modifiers: "Bold"),
-                                                off_style: (fg: "blue", modifiers: "Dim"),
-                                                oneshot_style: (fg: "red", modifiers: "Bold"),
-                                            )))),
-                                            (kind: Property(Status(ConsumeV2(
-                                                on_label: " ",
-                                                off_label: " ",
-                                                oneshot_label: "",
-                                                on_style: (fg: "yellow", modifiers: "Bold"),
-                                                off_style: (fg: "blue", modifiers: "Dim"),
-                                                oneshot_style: (fg: "red", modifiers: "Dim"),
-                                            )))),
-                                        ], align: Right)),
+                                        pane: Pane(Property(
+                                            format: "
+                                                $repeat(onLabel: ' ', offLabel: ' ', onStyle:  {fg: yellow, mods: bold}, offStyle: {fg: yellow, mods: dim})
+                                                $random(onLabel: ' ', offLabel: ' ', onStyle: {fg: yellow, mods: bold}, offStyle: {fg: yellow, mods: dim})
+                                                $single(onLabel: '󰑘 ', offLabel: '󰑘 ', oneshotLabel: '󰑘 ', onStyle: {fg: yellow, mods: bold}, offStyle: {fg: yellow, mods: dim}, oneshotStyle: {fg: red, mods: bold})
+                                                $consume(onLabel: ' ', offLabel: ' ', oneshotLabel: ' ', onStyle: {fg: yellow, mods: bold}, offStyle: {fg: yellow, mods: dim}, oneshotStyle: {fg: red, mods: dim})
+                                            ", align: Right)),
                                     ),
                                 ],
                             ),
@@ -147,9 +91,9 @@
                     direction: Horizontal,
                     panes: [
                         (
-                            pane: Pane(Property(content: [(kind: Property(Status(StateV2(playing_label: "  ", paused_label: "  ", stopped_label: "  ",
-                                playing_style: (fg: "blue"), paused_style: (fg: "green"), stopped_style: (fg: "red")
-                            ))))], align: Left)),
+                            pane: Pane(Property(
+                                format: "$state(playingLabel: '  ', pausedLabel: '  ', stoppedLabel: '  ', playingStyle: {fg: blue}, pausedStyle: {fg: green}, stoppedStyle: {fg: red})",
+                                align: Left)),
                             size: "3",
                         ),
                         (
@@ -158,13 +102,7 @@
                         ),
                         (
                             size: "13",
-                            pane: Pane(Property(
-                                content: [
-                                    (kind: Property(Status(Elapsed))),
-                                    (kind: Text(" / ")),
-                                    (kind: Property(Status(Duration))),
-                                ], align: Right,
-                            )),
+                            pane: Pane(Property(format: "$elapsed$' / '$duration", align: Right)),
                         ),
                     ]
                 ),
@@ -190,22 +128,7 @@
         thumb_style: (fg: "blue"),
     ),
     browser_column_widths: [20, 38, 42],
-    browser_song_format: [
-        (
-            kind: Group([
-                (kind: Property(Track)),
-                (kind: Text(" ")),
-            ])
-        ),
-        (
-            kind: Group([
-                (kind: Property(Artist)),
-                (kind: Text(" - ")),
-                (kind: Property(Title)),
-            ]),
-            default: (kind: Property(Filename))
-        ),
-    ],
+    browser_song_format_v2: "$[$s:track$' ']$[$s:artist$' - '$s:title]|$s:filename$'.'$s:fileextension",
     tab_bar: (
         active_style: (fg: "black", bg: "blue", modifiers: "Bold"),
         inactive_style: (),
@@ -215,25 +138,10 @@
     borders_style: (fg: "blue", modifiers: "Bold"),
     highlight_border_style: (fg: "red"),
     song_table_format: [
-        (
-            prop: (kind: Property(Other("albumartist")), default: (kind: Property(Artist), default: (kind: Text("Unknown")))),
-            width: "20%",
-            label: "Artist"
-        ),
-        (
-            prop: (kind: Property(Title), default: (kind: Text("Unknown"))),
-            width: "35%",
-        ),
-        (
-            prop: (kind: Property(Album), default: (kind: Text("Unknown Album"))),
-            width: "45%",
-        ),
-        (
-            prop: (kind: Property(Duration),default: (kind: Text("-"))),
-            width: "5",
-            alignment: Right,
-            label: "Len"
-        ),
+        (format: "$s:albumartist|$s:artist|$'unknown'", width: "20%", label: "Artist"),
+        (format: "$s:title|$'Unknown'", width: "35%"),
+        (format: "$s:album|$'Unknown Album'", width: "45%"),
+        (format: "$s:duration|$'-'", width: "5", alignment: Right, label: "Len"),
     ],
     header: (rows: []),
 )
```
</details>

roxwize commented:
I haven't been able to offer anything to this because I've been dealing with life etc. for a while. Got sick and had some personal issues and stuff so I don't know how long until I'll have anything to offer. I will say that the syntax is a bit strange but I'll get to that when I'm not feeling terrible 24/7.

Gri-ffin commented on src/config/tabs.rs:
```diff
@@ -148,20 +155,34 @@ impl From<PaneTypeFile> for PaneType {
+            PaneTypeFile::Property { content, align, scroll_speed, format } => PaneType::Property {
+                content: {
+                    match (format, content) {
+                        (Some(format), Some(_)) | (Some(format), None) => parser::parser()
+                            .parse(&format)
+                            .into_result()
+                            .map_err(|e| {
+                                anyhow::anyhow!("Failed to parse property format: {:?}", e)
+                            })?
+                            .into_iter()
+                            .map(|prop| -> Result<_> { prop.try_into() })
+                            .try_collect()?,
```
This should be documented, users should know that format will take precedence, or it might spark confusion

Gri-ffin commented on src/config/theme/parser.rs:
```diff
@@ -0,0 +1,891 @@
+
+    let escape_single_queoted_str =
+        choice((
+            just('\\').then(just('\'')).to('a').ignored(),
```
Couldn't be just be just('\\').then(just('\'')).to('\'').ignored(), forgive me if i'm wrong, but i don't understand what does the a do here?

Gri-ffin commented on src/config/theme/parser.rs:
```diff
@@ -0,0 +1,891 @@
+            just('t').to('\t'),
+            just('u').ignore_then(text::digits(16).exactly(4).to_slice().validate(
+                |digits, e, emitter| {
+                    char::from_u32(
+                        u32::from_str_radix(digits, 16)
+                            .expect("Only valid digits should have been parsed in text::digits"),
+                    )
```
Wouldn't it be better to use chumsky here too? wouldn't this be hell to debug as a user? or is there a reason to crash here?

Gri-ffin commented on src/config/theme/parser.rs:
```diff
@@ -0,0 +1,891 @@
+            just('r').to('\r'),
+            just('t').to('\t'),
+            just('u').ignore_then(text::digits(16).exactly(4).to_slice().validate(
+                |digits, e, emitter| {
+                    char::from_u32(
+                        u32::from_str_radix(digits, 16)
+                            .expect("Only valid digits should have been parsed in text::digits"),
+                    )
```
Wouldn't it be better to use chumsky here too and exit gracefully? wouldn't this be hell to debug as a user? or is there a reason to crash here?

mierak replied:
The thing is, this invariant should be ensured by this `text::digits(16)` a few lines above. 

This panic should never really happen and if it does it should get fixed asap. But I suppose a more graceful error is fine too here.

Gri-ffin commented on src/config/theme/parser.rs:
```diff
@@ -0,0 +1,891 @@
+                    just('t').to('\t'),
+                    just('u').ignore_then(text::digits(16).exactly(4).to_slice().validate(
+                        |digits, e, emitter| {
+                            char::from_u32(u32::from_str_radix(digits, 16).expect(
+                                "Only valid digits should have been parsed in text::digits",
+                            ))
```
Same thing, here i think its best to use chumsky for error reporting and exit gracefully

Gri-ffin commented on src/config/theme/queue_table.rs:
```diff
@@ -160,7 +168,20 @@ impl TryFrom<QueueTableColumnsFile> for QueueTableColumns {
+                            .context("A song property must be specified")?
+                            .try_into()?,
+                        (None, Some(prop)) => prop.try_into()?,
+                        (None, None) => panic!(""),
```
this would be a nightmare to debug as a user no? bailing would be better or atleast a panic with a message?

mierak replied:
Thats just a left over from quickly getting this to work, take the PR with a grain of salt, its by no means close to finished.

mierak commented:
I introduced some shorter variants for the properties, for example "artist" and "ar" are equivalent. I have also decided to namespace the properties to avoid current and future conflicts.

For string, $ is no longer needed, simply surrounding text in double or single quotes is fine, same for groups, no dollar sign needed, only brackets.

Introduced the truncate transform as either `%trunc` or `%truncate`.

Attributes on properties are now `snake_case` to be inline with the rest of the config.

I have also decided to namespace the properties, in the same way they are now with enums to avoid conflicts - still undecided on this one, opinions anyone?

And some general cleanup. Whats left is mostly:

* Docs - thats probably bulk of the work
* Tests - some are there but not nearly enough
* finalize config deserialization and conversion
* Cleanups and refactors of the code

mierak commented:
So this is basically done but hit a bit of a roadblock https://github.com/ron-rs/ron/issues/577

The current implementation with this issue will break existing configs if they use stuff like `Property(Status(StateV2()))` but `Property(Status(StateV2( )))`(not the space between parens) is fine.

Either we introduce a new set of options (browser_song_format_v2, format instead of content for Property panes, etc) like it was in the first place, or we wait whether this will get fixed at some point. I dont see any other options.

mierak commented:
> What would the disadvantage be if we introduce a new set of options? I am not sure I understand what you mean with this

If the issue I reported didnt exist, then we could have `browser_song_format` accept both `browser_song_format: "$file"` and `browser_song_format: [(kind: Property(Filename))]`. But without it getting fixed we would need `browser_song_format` and `browser_song_format_v2` for one or the other or accept that existing config files will get broken.

Same goes for the other places: song table format, property panes, tabs and layout

mierak commented:
> temporally adding

I personally would not mind, but once something is added and used by anyone it will then be a breaking change to their config to remove the option again and I want to avoid those as much as possible.
